### PR TITLE
Correctly setup `MockQuery.snapshots()` to broadcast data to 'already listening' streams.

### PR DESCRIPTION
**Pull request by gki**
# Objective
Fix the issue #111 

# Changes
- Add `QuerySnapshotStreamManager` to keep and re-fire updated data to streams for each `Query`.
- Setup `snapshots()` stream to not broadcast same data.
- Add unit test

# Tests
- [x] listening snapshot stream with `where`.
- [x] listening snapshot stream with `orderBy`.
- [x] listening snapshot stream with `where` and `orderBy`.

### CODE REVIEW

**Comment by gki** [comment]
Not designed well. close for now.

**Comment by atn832** [comment]
Thanks for filing and working on it, @gki! I'll take a look at your draft in the next few days.

**Comment by gki** [comment]
Thanks @atn832 . Actually, I'm not familiar to sending PR to OSS repo.
So feel free to point out my code if there are unsuitable structure, naming, updates etc, etc... 😉

**Comment by atn832** [comment]
Awesome work, I only have minor comments.

File: lib/src/mock_query.dart [hunk 1]
```diff
@@ -197,3 +224,76 @@ class MockQuery extends Mock implements Query {
...
+      {};
+
+  void clear() {
+    for (var maps in _streamCache.values) {
+      maps.forEach((_, value) => value.close());
+    }
+    _streamCache.clear();
+  }
+
+  String _retribeParentPath(MockQuery query) {
```
[the remark this answers is not in the document]
Comment (gki) [reply lib/src/mock_query.dart]: done

File: lib/src/mock_query.dart [hunk 2]
```diff
@@ -197,3 +224,76 @@ class MockQuery extends Mock implements Query {
...
+      _instance ??= QuerySnapshotStreamManager._internal();
+
+  QuerySnapshotStreamManager._internal();
+  final Map<String, Map<Query, StreamController<QuerySnapshot>>> _streamCache =
+      {};
+
+  void clear() {
+    for (var maps in _streamCache.values) {
+      maps.forEach((_, value) => value.close());
+    }
```
[the remark this answers is not in the document]
Comment (gki) [reply lib/src/mock_query.dart]: done. Refactored to the suggested code.

File: lib/src/mock_query.dart [hunk 3]
```diff
@@ -197,3 +224,76 @@ class MockQuery extends Mock implements Query {
...
+    return pathCache[query];
+  }
+
+  void fireSnapshotUpdate(String path) {
+    final pathCache = _streamCache[path];
+    if (pathCache == null) {
+      return;
+    }
+    // print('fireSnapshotUpdate cache (for $path) length ${pathCache.length}');
```
[the remark this answers is not in the document]
Comment (gki) [reply lib/src/mock_query.dart:289]: done

File: lib/src/mock_query.dart [hunk 4]
```diff
@@ -197,3 +224,76 @@ class MockQuery extends Mock implements Query {
...
+    final noListnerQueries = <Query>[];
```
[the remark this answers is not in the document]
Comment (gki) [reply lib/src/mock_query.dart]: done

File: test/mock_query_test.dart [hunk 5]
```diff
@@ -508,4 +505,201 @@ void main() {
...
+    });
+
+    // check new stream will receive the latest data.
+    instance
+        .collection('messages')
+        .where('archived', isEqualTo: false)
+        .snapshots()
+        .listen(expectAsync1((snapshot) {
+      expect(snapshot.documents.length, equals(2));
+      for (var d in snapshot.documents) {
```
[the remark this answers is not in the document]
Comment (gki) [reply test/mock_query_test.dart]: done

File: lib/src/mock_query.dart [hunk 6]
```diff
@@ -36,9 +42,30 @@ class MockQuery extends Mock implements Query {
     return MockSnapshot(documents);
   }
 
+  final _unOrdDeepEq = const DeepCollectionEquality.unordered();
```
[the remark this answers is not in the document]
Comment (gki) [reply lib/src/mock_query.dart]: done

File: test/mock_query_test.dart [hunk 7]
```diff
@@ -508,4 +505,201 @@ void main() {
...
+        'content': 'bonjour!',
+        'receivedAt': now.add(const Duration(seconds: 1)),
+      },
+      {
+        'content': 'hola!',
+        'receivedAt': now.subtract(const Duration(seconds: 1)),
+      }
+    ];
+
+    final ascContnts = [
```
[the remark this answers is not in the document]
Comment (gki) [reply test/mock_query_test.dart]: done

File: test/mock_query_test.dart [hunk 8]
```diff
@@ -508,4 +505,201 @@ void main() {
...
+      ['hello!'],
+      ['hello!', 'bonjour!'],
+      ['hola!', 'hello!', 'bonjour!'],
+    ];
+
+    final descContnts = [
```
[the remark this answers is not in the document]
Comment (gki) [reply test/mock_query_test.dart]: done

File: lib/src/mock_query.dart [hunk 9]
```diff
@@ -197,3 +224,76 @@ class MockQuery extends Mock implements Query {
...
+  }
+
+  void fireSnapshotUpdate(String path) {
+    final pathCache = _streamCache[path];
+    if (pathCache == null) {
+      return;
+    }
+    // print('fireSnapshotUpdate cache (for $path) length ${pathCache.length}');
+    final noListnerQueries = <Query>[];
+    for (final query in pathCache.keys) {
```
[the remark this answers is not in the document]
Comment (gki) [reply lib/src/mock_query.dart:290]: Thanks. 

> To be honest, I don't mind not cleaning streams without listeners unless someone runs into actual memory issues. 

I agree with your thought. Maybe I worried too much.
I've deleted the code to remove `queriesWithNoListener` from `pathCache`.

File: lib/src/mock_query.dart [hunk 10]
```diff
@@ -36,9 +42,30 @@ class MockQuery extends Mock implements Query {
...
+          return false;
+        }
+
+        if (!_unOrdDeepEq.equals(
+            prev.documents[i].data, next.documents[i].data)) {
+          return false;
+        }
+      }
+      return true;
+    });
```
[the remark this answers is not in the document]
Comment (gki) [reply lib/src/mock_query.dart]: done

**Comment by gki** [comment]
@atn832 Thanks for reviewing!

File: lib/src/mock_query.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (atn832) [reply lib/src/mock_query.dart]: It should be retrieve, not retrive. :) But it's ok, I'll merge it now, and we can fix the typo later.

File: test/mock_query_test.dart [hunk 11]
```diff
@@ -508,4 +505,201 @@ void main() {
...
+    instance
+        .collection('messages')
+        .where('archived', isEqualTo: true)
+        .snapshots()
+        .listen(expectAsync1((snapshot) {
+          expect(snapshot.documents.length, inInclusiveRange(0, 1));
+          for (var d in snapshot.documents) {
+            expect(d.data['archived'], isTrue);
+          }
+        }, count: 2)); // initial [], when add 'hello!' and when add 'hola!'.
```
[the remark this answers is not in the document]
Comment (atn832) [reply test/mock_query_test.dart:531]: Did you forget to update the comment here?

**Comment by atn832** [comment]
Thank you for this big, well-written PR! I'll publish an update later this week.

**Comment by gki** [comment]
Ah, sorry to incomplete update... 😩 I will create another PR to fix remaining points.

**Comment by atn832** [comment]
I just published this and #115 to https://pub.dev/packages/cloud_firestore_mocks/changelog#0461. Thank you for your awesome contributions!

**Pull request merged**